backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inout/ShipmentScheduleQtyReservedStorage.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * #%L
+ * de.metas.handlingunits.base
+ * %%
+ * Copyright (C) 2023 metas GmbH
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
+package de.metas.handlingunits.inout;
+
+import com.google.common.collect.ImmutableList;
+import de.metas.cache.CCache;
+import de.metas.inoutcandidate.api.IShipmentScheduleEffectiveBL;
+import de.metas.inoutcandidate.model.I_M_ShipmentSchedule;
+import de.metas.material.cockpit.stock.StockDataQuery;
+import de.metas.material.event.commons.AttributesKey;
+import de.metas.product.ProductId;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.adempiere.inout.util.IShipmentScheduleQtyOnHandStorage;
+import org.adempiere.inout.util.ShipmentScheduleAvailableStockDetail;
+import org.adempiere.util.lang.impl.TableRecordReference;
+import org.adempiere.warehouse.WarehouseId;
+import org.adempiere.warehouse.api.IWarehouseDAO;
+import org.compiere.util.Util;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * This class shall hold details of reserved stock that's ready to be shipped (= VHUs that are reserved for one or more OrderLine).
+ */
+public class ShipmentScheduleQtyReservedStorage implements IShipmentScheduleQtyOnHandStorage
+{
+	private final CCache<Util.ArrayKey, StockDataQuery> cachedMaterialQueries = CCache.newLRUCache("QtyReservedCachedMaterialQueries", 200, CCache.EXPIREMINUTES_Never);
+	private final IShipmentScheduleEffectiveBL shipmentScheduleEffectiveBL = Services.get(IShipmentScheduleEffectiveBL.class);
+	private final IWarehouseDAO warehouseDAO = Services.get(IWarehouseDAO.class);
+	@NonNull final ImmutableList<ShipmentScheduleAvailableStockDetail> stockDetails;
+
+	public ShipmentScheduleQtyReservedStorage(final @NonNull ImmutableList<ShipmentScheduleAvailableStockDetail> stockDetails)
+	{
+		this.stockDetails = stockDetails;
+	}
+
+	@Override
+	public List<ShipmentScheduleAvailableStockDetail> getStockDetailsMatching(final @NonNull I_M_ShipmentSchedule sched)
+	{
+		if (stockDetails.isEmpty())
+		{
+			return Collections.emptyList();
+		}
+		else
+		{
+			//
+			// Reserved qty shall always be in VHUs, no need for picking BOM.
+			final StockDataQuery mainProductQuery = toQuery(sched);
+			return new ArrayList<>(getStockDetailsMatching(mainProductQuery));
+		}
+	}
+
```

Comment:
just a suggestion, generally I like to use @NonNull/@Nullable on the function result as well as the params :)

Reply:
done.